# NUMPY TASK


**SMART CITY TRAFFIC & RETAIL ANALYTICS**

#  SCENARIO

**You are working as a Data Analyst in a Smart City Analytics Project.**
- The city collects:
- Traffic data


- Retail sales data


- Weather data


- Your task is to analyze these datasets using ONLY NumPy


# DATA GENERATION


**A. Traffic Dataset**
- Traffic data collected for:
- 4 regions


- 7 days


- 24 hours
- Shape → (4, 7, 24)
- import numpy as np
- np.random.seed(10)


- traffic = np.random.randint(50, 500, (4,7,24))

- Each value = number of vehicles per hour




In [1]:
import numpy as np
np.random.seed(10)


traffic = np.random.randint(50, 500, (4,7,24))
traffic

array([[[315, 175,  65, 370, 419, 173, 206, 271, 463, 419,  58, 123,
         306,  90, 470, 421,  66, 406, 289, 445, 104, 394, 413, 172],
        [112,  83, 250, 406, 227, 229, 104, 127,  63, 293, 459, 432,
         191, 398, 392, 336, 208, 139, 190, 371,  81, 363, 342, 333],
        [ 68, 143, 127, 456, 457, 144, 298, 445, 206, 124, 266, 443,
          65,  68, 168, 121, 394, 445, 195, 480, 185, 253, 334, 339],
        [300, 402,  94, 183,  54, 121, 394, 394, 356, 414, 232, 340,
         410, 321, 394, 193, 184, 296, 135, 284, 200,  61, 162, 318],
        [270, 112, 235, 129, 414, 282,  92, 287, 363, 285, 403, 356,
          95, 474, 251, 215, 178, 452,  73,  53, 335, 322, 390, 388],
        [162,  64, 101, 129, 293, 323, 228, 103, 459, 416, 482, 284,
         323, 466, 387, 283,  91, 140, 318, 336, 259, 451, 194, 178],
        [209, 407, 251, 498, 294, 430, 472, 304,  72, 146, 372, 117,
         496, 145, 405, 461, 132, 112, 296, 127, 413,  98, 423, 287]],

       [[165, 471, 317, 4

**B. Sales Dataset**
- Retail sales data for:
- 12 months


- 5 stores


- 8 products


- Shape → (12, 5, 8)
- sales = np.random.randint(100, 1000, (12,5,8))



In [2]:
sales = np.random.randint(100, 1000, (12,5,8))
sales

array([[[737, 936, 230, 627, 555, 197, 974, 663],
        [681, 264, 814, 128, 949, 411, 300, 490],
        [372, 995, 888, 615, 532, 560, 688, 286],
        [165, 694, 891, 636, 556, 410, 727, 121],
        [794, 759, 451, 410, 103, 723, 710, 509]],

       [[781, 114, 813, 683, 266, 533, 890, 724],
        [709, 147, 259, 914, 198, 489, 498, 944],
        [372, 320, 632, 566, 365, 924, 777, 710],
        [362, 770, 679, 298, 409, 284, 369, 479],
        [878, 672, 739, 738, 645, 228, 572, 668]],

       [[776, 978, 934, 602, 714, 617, 283, 869],
        [813, 535, 241, 619, 609, 894, 622, 600],
        [850, 354, 266, 988, 365, 966, 798, 773],
        [890, 958, 428, 538, 679, 396, 866, 787],
        [843, 870, 208, 635, 337, 993, 515, 933]],

       [[668, 904, 699, 744, 690, 199, 415, 826],
        [386, 715, 872, 827, 393, 978, 867, 934],
        [671, 647, 678, 797, 286, 868, 792, 235],
        [222, 769, 142, 854, 556, 820, 854, 619],
        [260, 184, 368, 692, 857, 714, 200, 

**C. Weather Dataset**
- Temperature for 365 days
- Shape → (365,)
- temperature = np.random.normal(30, 5, 365)


In [3]:
temperature = np.random.normal(30, 5, 365)
print(temperature)
print(temperature.shape)

[37.67447168 33.43330212 32.05288001 17.74729097 28.1582116  32.92125768
 25.86234172 34.28335871 33.00901823 31.68290842 30.52398537 38.4543455
 32.09553607 26.66732852 38.07631572 35.18099408 30.21298891 24.31127611
 27.24651875 34.67691917 27.99717021 33.69876873 20.09455489 25.29691485
 32.27376729 28.31301242 24.45504814 28.38081852 37.7843414  37.94854595
 34.00821496 30.15730418 33.17452057 31.60982997 30.02879826 32.24042004
 29.90123055 33.67518868 27.99852975 29.76047734 31.9574244  26.51433221
 25.10837905 29.21509114 31.69788532 24.37602704 30.23461545 31.69926106
 30.31684261 24.94072962 31.04136861 28.0229218  31.7364542  27.27120278
 16.60015522 24.60582144 33.15878614 23.85524207 32.0626499  23.34533122
 26.76777513 26.38625535 25.36063318 29.65298342 30.99677909 22.97164746
 28.60839783 27.95919581 27.45092449 36.24849357 35.92097384 33.53780572
 27.2553888  33.09281413 32.65063038 31.78193144 28.73049932 20.59979979
 31.31129964 29.08224724 30.37538785 27.40572144 34.

**Task 1 – Explore Properties**
- Find for both datasets:
- ndim


- shape


- size


- dtype




In [12]:
print("first datset:")
print("dimensions of traffic dataset:",traffic.ndim)
print("shape  of traffic dataset:",traffic.shape)
print("size of traffic dataset:",traffic.size)
print("dtype of traffic dataset:",traffic.dtype)

print("second data set:")
print("dimensions of sales dataset:",sales.ndim)
print("shape  of sales dataset:",sales.shape)
print("size of sales dataset:",sales.size)
print("dtype of sales dataset:",sales.dtype)

first datset:
dimensions of traffic dataset: 3
shape  of traffic dataset: (4, 7, 24)
size of traffic dataset: 672
dtype of traffic dataset: int32
second data set:
dimensions of sales dataset: 3
shape  of sales dataset: (12, 5, 8)
size of sales dataset: 480
dtype of sales dataset: int32


**Task 2 – Reshaping Practice**
- Without changing data:
- Reshape traffic data to (28, 24)


- Reshape sales data to (60, 8)




In [13]:
reshape=traffic.reshape(28,24)
reshape_sales=sales.reshape(60,8)
print("reshape traffic data to (28,24)\n:",reshape)
print("reshape sales data to (60,8)\n:",reshape_sales)

reshape traffic data to (28,24)
: [[315 175  65 370 419 173 206 271 463 419  58 123 306  90 470 421  66 406
  289 445 104 394 413 172]
 [112  83 250 406 227 229 104 127  63 293 459 432 191 398 392 336 208 139
  190 371  81 363 342 333]
 [ 68 143 127 456 457 144 298 445 206 124 266 443  65  68 168 121 394 445
  195 480 185 253 334 339]
 [300 402  94 183  54 121 394 394 356 414 232 340 410 321 394 193 184 296
  135 284 200  61 162 318]
 [270 112 235 129 414 282  92 287 363 285 403 356  95 474 251 215 178 452
   73  53 335 322 390 388]
 [162  64 101 129 293 323 228 103 459 416 482 284 323 466 387 283  91 140
  318 336 259 451 194 178]
 [209 407 251 498 294 430 472 304  72 146 372 117 496 145 405 461 132 112
  296 127 413  98 423 287]
 [165 471 317 455 467 145 221 274 279 123 218 407 477 285 396 121 186 263
  378 462 305 415  80 173]
 [203 256 263 496  63 219 467  54 265 400  78  89 397 153 315 313 200  82
  309  59 230 246 208 426]
 [422 380 336 187 308 499 165 191 486 183 271 390 482 496

**Task 3 – Flattening Concept**
- Flatten traffic dataset


- Flatten sales dataset


- Compare behavior of:
- flatten()
- ravel()


In [5]:
fal=traffic.flatten()
fla_sales=sales.flatten()
ravel=traffic.ravel()
ravel_sales=sales.ravel()
print("Flatten traffic dataset\n:",fal)
print("flatten sales dataset\n:",fla_sales)
print("ravel traffic dataset\n:",ravel)
print("ravel sales dataset\n:",ravel_sales)

Flatten traffic dataset
: [315 175  65 370 419 173 206 271 463 419  58 123 306  90 470 421  66 406
 289 445 104 394 413 172 112  83 250 406 227 229 104 127  63 293 459 432
 191 398 392 336 208 139 190 371  81 363 342 333  68 143 127 456 457 144
 298 445 206 124 266 443  65  68 168 121 394 445 195 480 185 253 334 339
 300 402  94 183  54 121 394 394 356 414 232 340 410 321 394 193 184 296
 135 284 200  61 162 318 270 112 235 129 414 282  92 287 363 285 403 356
  95 474 251 215 178 452  73  53 335 322 390 388 162  64 101 129 293 323
 228 103 459 416 482 284 323 466 387 283  91 140 318 336 259 451 194 178
 209 407 251 498 294 430 472 304  72 146 372 117 496 145 405 461 132 112
 296 127 413  98 423 287 165 471 317 455 467 145 221 274 279 123 218 407
 477 285 396 121 186 263 378 462 305 415  80 173 203 256 263 496  63 219
 467  54 265 400  78  89 397 153 315 313 200  82 309  59 230 246 208 426
 422 380 336 187 308 499 165 191 486 183 271 390 482 496 348 212 218  96
 466 400 136 236 426 351 

**flatten():**

- Returns a new copy of all 480 elements in 1D.

- Shape (480,).

- Safe to modify without touching the original sales dataset.

**ravel():**

- Returns a view of the same 480 elements in 1D.

- Shape (480,).

- Modifications propagate back to the original sales dataset.

**Task 4 – 2D Style Slicing from Sales Data**
- Extract:
- Data of store 2 for all months


- Data of product 4 for all stores and months


- First 6 months and first 3 products


- Last 4 months of sales




In [22]:
#Data of store 2 for all months
slicing=sales[:,1,:]
print("Data of store 2 for all months\n:",slicing)

Data of store 2 for all months
: [[681 264 814 128 949 411 300 490]
 [709 147 259 914 198 489 498 944]
 [813 535 241 619 609 894 622 600]
 [386 715 872 827 393 978 867 934]
 [708 285 892 363 951 585 397 119]
 [931 752 702 158 558 352 802 270]
 [416 453 292 968 841 166 108 983]
 [112 105 971 684 796 982 649 393]
 [857 263 409 235 812 925 192 674]
 [408 316 782 936 457 477 232 442]
 [343 685 199 350 216 613 449 882]
 [255 752 592 776 654 909 541 384]]


In [24]:
#Data of product 4 for all stores and months
slicing_2=sales[:,:,3]
print("Data of product 4 for all stores and months\n:",slicing_2)

Data of product 4 for all stores and months
: [[627 128 615 636 410]
 [683 914 566 298 738]
 [602 619 988 538 635]
 [744 827 797 854 692]
 [523 363 216 764 729]
 [422 158 907 648 722]
 [576 968 121 728 437]
 [775 684 362 709 742]
 [358 235 123 756 119]
 [949 936 306 543 615]
 [486 350 188 790 240]
 [734 776 357 392 649]]


In [16]:
#First 6 months and first 3 products
slicing_3=sales[:6,:,:3]
print("First 6 months and first 3 products\n:",slicing_3)

First 6 months and first 3 products
: [[[737 936 230]
  [681 264 814]
  [372 995 888]
  [165 694 891]
  [794 759 451]]

 [[781 114 813]
  [709 147 259]
  [372 320 632]
  [362 770 679]
  [878 672 739]]

 [[776 978 934]
  [813 535 241]
  [850 354 266]
  [890 958 428]
  [843 870 208]]

 [[668 904 699]
  [386 715 872]
  [671 647 678]
  [222 769 142]
  [260 184 368]]

 [[649 684 337]
  [708 285 892]
  [731 115 314]
  [557 498 393]
  [947 809 451]]

 [[194 967 858]
  [931 752 702]
  [138 786 630]
  [923 596 685]
  [331 112 964]]]


In [25]:
#Last 4 months of sales
s4=sales[-4:,:,:]
print("Last 4 months of sales\n:",s4)

Last 4 months of sales
: [[[256 787 631 358 727 548 551 814]
  [857 263 409 235 812 925 192 674]
  [287 912 666 123 746 238 984 297]
  [505 385 157 756 574 200 338 247]
  [730 331 916 119 680 968 608 241]]

 [[249 783 723 949 578 973 295 910]
  [408 316 782 936 457 477 232 442]
  [663 153 515 306 427 485 337 626]
  [845 811 145 543 482 534 216 730]
  [410 604 393 615 293 376 110 162]]

 [[711 154 887 486 759 835 553 588]
  [343 685 199 350 216 613 449 882]
  [836 108 342 188 744 568 493 687]
  [305 386 349 790 353 402 179 510]
  [859 682 283 240 250 848 315 201]]

 [[507 787 605 734 298 577 235 159]
  [255 752 592 776 654 909 541 384]
  [195 499 336 357 765 335 259 690]
  [745 626 727 392 579 958 388 386]
  [606 758 340 649 893 449 362 888]]]


**Task 5 – 3D Slicing from Traffic Data**
- Extract:
- Region 1 data for all days


- Data for day 3 across all regions


- Traffic from 8 AM to 5 PM only


- Traffic of last region and last day



In [10]:
#Region 1 data for all days
t1=traffic[0,:,:]
print("Region 1 Data for all days\n:",t1)

Region 1 Data for all days
: [[315 175  65 370 419 173 206 271 463 419  58 123 306  90 470 421  66 406
  289 445 104 394 413 172]
 [112  83 250 406 227 229 104 127  63 293 459 432 191 398 392 336 208 139
  190 371  81 363 342 333]
 [ 68 143 127 456 457 144 298 445 206 124 266 443  65  68 168 121 394 445
  195 480 185 253 334 339]
 [300 402  94 183  54 121 394 394 356 414 232 340 410 321 394 193 184 296
  135 284 200  61 162 318]
 [270 112 235 129 414 282  92 287 363 285 403 356  95 474 251 215 178 452
   73  53 335 322 390 388]
 [162  64 101 129 293 323 228 103 459 416 482 284 323 466 387 283  91 140
  318 336 259 451 194 178]
 [209 407 251 498 294 430 472 304  72 146 372 117 496 145 405 461 132 112
  296 127 413  98 423 287]]


In [11]:
#Data for day 3 across all regions
t2=traffic[:,2,:]
print("Data for day 3 across all regions\n:",t2)

Data for day 3 across all regions
: [[ 68 143 127 456 457 144 298 445 206 124 266 443  65  68 168 121 394 445
  195 480 185 253 334 339]
 [422 380 336 187 308 499 165 191 486 183 271 390 482 496 348 212 218  96
  466 400 136 236 426 351]
 [113 389 381 220 427 472 235 185 274 411  88 416  98 205 415 370 227 315
  333 358 396 363 482 163]
 [282 204 386 168 462 204 400 179 397 436 266 366 485 191 383 331 239 474
  440 409 161 471 131 120]]


In [27]:
#Traffic from 8 AM to 5 PM only
t3=traffic[:,:,7:17]
print("Traffic from 8 AM to 5 PM only\n:",t3)

Traffic from 8 AM to 5 PM only
: [[[271 463 419  58 123 306  90 470 421  66]
  [127  63 293 459 432 191 398 392 336 208]
  [445 206 124 266 443  65  68 168 121 394]
  [394 356 414 232 340 410 321 394 193 184]
  [287 363 285 403 356  95 474 251 215 178]
  [103 459 416 482 284 323 466 387 283  91]
  [304  72 146 372 117 496 145 405 461 132]]

 [[274 279 123 218 407 477 285 396 121 186]
  [ 54 265 400  78  89 397 153 315 313 200]
  [191 486 183 271 390 482 496 348 212 218]
  [ 68 461 133 215 153 193 105 440 350  59]
  [454 443  90 194 381 473 364 259  89  99]
  [195 463 228 101 198  55 297 414 248 239]
  [360  69 347 239 444  86  50 301 208 368]]

 [[ 70 465 139 174 306 402 331 113 365 308]
  [476 137 363 300 258  52 212 108 188 129]
  [185 274 411  88 416  98 205 415 370 227]
  [125 143 336 337 206 151 262 199 262 277]
  [ 87 209 362 259 393 432 105 403 124 298]
  [232 149 388 234  91 396  76  58 210 377]
  [485 356 228  64 426 486 464 131  60 367]]

 [[214 440 113 297 366  53 356 109 29

In [13]:
#Traffic of last region and last day
t5=traffic[-1,-1,:]
print("Traffic of last region and last day\n:",t5)

Traffic of last region and last day
: [394 427 220  71 265 160 452 325 279 277 210 201 265 481 362 353 340 274
 405 323 402 146 417 452]


**Task 6 – Reverse Slicing**
- Reverse order of months in sales


- Reverse order of hours in traffic


In [14]:
#Reverse order of months in sales
reverse=sales[::-1,:,:]
print("Reverse order of months in sales\n:",reverse)

Reverse order of months in sales
: [[[507 787 605 734 298 577 235 159]
  [255 752 592 776 654 909 541 384]
  [195 499 336 357 765 335 259 690]
  [745 626 727 392 579 958 388 386]
  [606 758 340 649 893 449 362 888]]

 [[711 154 887 486 759 835 553 588]
  [343 685 199 350 216 613 449 882]
  [836 108 342 188 744 568 493 687]
  [305 386 349 790 353 402 179 510]
  [859 682 283 240 250 848 315 201]]

 [[249 783 723 949 578 973 295 910]
  [408 316 782 936 457 477 232 442]
  [663 153 515 306 427 485 337 626]
  [845 811 145 543 482 534 216 730]
  [410 604 393 615 293 376 110 162]]

 [[256 787 631 358 727 548 551 814]
  [857 263 409 235 812 925 192 674]
  [287 912 666 123 746 238 984 297]
  [505 385 157 756 574 200 338 247]
  [730 331 916 119 680 968 608 241]]

 [[646 816 813 775 697 811 237 897]
  [112 105 971 684 796 982 649 393]
  [397 346 392 362 377 149 165 281]
  [415 486 529 709 584 321 214 276]
  [402 316 904 742 557 460 565 392]]

 [[660 587 217 576 394 806 544 730]
  [416 453 292 968 

In [15]:
#Reverse order of hours in traffic
reverse_hour=traffic[:,:,::-1]
print("Reverse order of hours in traffic\n:",reverse_hour)

Reverse order of hours in traffic
: [[[172 413 394 104 445 289 406  66 421 470  90 306 123  58 419 463 271
   206 173 419 370  65 175 315]
  [333 342 363  81 371 190 139 208 336 392 398 191 432 459 293  63 127
   104 229 227 406 250  83 112]
  [339 334 253 185 480 195 445 394 121 168  68  65 443 266 124 206 445
   298 144 457 456 127 143  68]
  [318 162  61 200 284 135 296 184 193 394 321 410 340 232 414 356 394
   394 121  54 183  94 402 300]
  [388 390 322 335  53  73 452 178 215 251 474  95 356 403 285 363 287
    92 282 414 129 235 112 270]
  [178 194 451 259 336 318 140  91 283 387 466 323 284 482 416 459 103
   228 323 293 129 101  64 162]
  [287 423  98 413 127 296 112 132 461 405 145 496 117 372 146  72 304
   472 430 294 498 251 407 209]]

 [[173  80 415 305 462 378 263 186 121 396 285 477 407 218 123 279 274
   221 145 467 455 317 471 165]
  [426 208 246 230  59 309  82 200 313 315 153 397  89  78 400 265  54
   467 219  63 496 263 256 203]
  [351 426 236 136 400 466  96 218 

**Task 7 – Traffic Aggregations**
-  Find:
- Total traffic per region


- Total traffic per day


- Total traffic per hour


In [16]:
#Total traffic per region
per_region=np.sum(traffic,axis=(1,2))
per_region

array([45119, 46828, 44283, 48317])

In [17]:
#Total traffic per day
per_day=np.sum(traffic,axis=(0,2))
per_day

array([26032, 23884, 28830, 27273, 26364, 25610, 26554])

In [18]:
#Total traffic per hour
per_hour=np.sum(traffic,axis=(0,1))
per_hour

array([7522, 8051, 7721, 7352, 8046, 7525, 8104, 7214, 8349, 7538, 6845,
       8167, 8219, 7045, 8017, 6929, 6874, 7295, 8514, 7530, 8002, 7849,
       8151, 7688])

**Task 8 – Sales Aggregations**
- Find:
- Total yearly sales


- Sales per month


- Sales per store


- Sales per product




In [19]:
#Total yearly sales
print("Total yearly sales:",np.sum(sales))

Total yearly sales: 261064


In [20]:
#Sales per month
print("Sales per month:",np.sum(sales,axis=(1,2)))

Sales per month: [22551 22418 26942 24623 20190 21356 19346 20725 21047 20291 19628 21947]


In [21]:
#Sales per store
print("Sales per store:",np.sum(sales,axis=(0,2)))

Sales per store: [58503 53861 48780 48024 51896]


In [22]:
#Sales per product
print("Sales per product:\n",np.sum(sales,axis=(0,1)))

Sales per product:
 [33211 34033 33255 34367 31904 33341 28967 31986]


**Task 9 – Averages**
- Compute:
- Average hourly traffic


- Average monthly sales


- Average sales per product




In [23]:
#Average hourly traffic
np.mean(traffic,axis=(0,1))

array([268.64285714, 287.53571429, 275.75      , 262.57142857,
       287.35714286, 268.75      , 289.42857143, 257.64285714,
       298.17857143, 269.21428571, 244.46428571, 291.67857143,
       293.53571429, 251.60714286, 286.32142857, 247.46428571,
       245.5       , 260.53571429, 304.07142857, 268.92857143,
       285.78571429, 280.32142857, 291.10714286, 274.57142857])

In [24]:
#Average monthly sales
np.mean(sales,axis=(1,2))

array([563.775, 560.45 , 673.55 , 615.575, 504.75 , 533.9  , 483.65 ,
       518.125, 526.175, 507.275, 490.7  , 548.675])

In [25]:
#Average sales per product
np.mean(sales,axis=(1,0))

array([553.51666667, 567.21666667, 554.25      , 572.78333333,
       531.73333333, 555.68333333, 482.78333333, 533.1       ])

**Task 10 – Statistical Measures**
- Compute for both datasets:
- mean


- median


- standard deviation


- variance


- min


- max


In [26]:
#compute traffic datset
print("traffic mean value:\n",np.mean(traffic))
print("traffic median value:\n",np.median(traffic))
print("traffic standard deviation value:\n",np.std(traffic))
print("traffic variance value:\n",np.var(traffic))
print("traffic min value:\n",np.min(traffic))
print("traffic max value:\n",np.max(traffic))

traffic mean value:
 274.6235119047619
traffic median value:
 278.5
traffic standard deviation value:
 130.97143766067708
traffic variance value:
 17153.517482904623
traffic min value:
 50
traffic max value:
 499


In [27]:
#compute sales datset
print("sales mean value:\n",np.mean(sales))
print("sales median value:\n",np.median(sales))
print("sales standard deviation value:\n",np.std(sales))
print("sales variance value:\n",np.var(sales))
print("sales min value:\n",np.min(sales))
print("sales max value:\n",np.max(sales))

sales mean value:
 543.8833333333333
sales median value:
 557.5
sales standard deviation value:
 251.48500397085752
sales variance value:
 63244.70722222223
sales min value:
 103
sales max value:
 995


**Task 11**
- Perform WITHOUT any loops:
- Increase all sales values by 15%


- Convert traffic values into “thousands of vehicles”


- Normalize temperature between 0 and 1



In [33]:
#Increase all sales values by 15%
sales*1.15

array([[[ 847.55, 1076.4 ,  264.5 ,  721.05,  638.25,  226.55, 1120.1 ,
          762.45],
        [ 783.15,  303.6 ,  936.1 ,  147.2 , 1091.35,  472.65,  345.  ,
          563.5 ],
        [ 427.8 , 1144.25, 1021.2 ,  707.25,  611.8 ,  644.  ,  791.2 ,
          328.9 ],
        [ 189.75,  798.1 , 1024.65,  731.4 ,  639.4 ,  471.5 ,  836.05,
          139.15],
        [ 913.1 ,  872.85,  518.65,  471.5 ,  118.45,  831.45,  816.5 ,
          585.35]],

       [[ 898.15,  131.1 ,  934.95,  785.45,  305.9 ,  612.95, 1023.5 ,
          832.6 ],
        [ 815.35,  169.05,  297.85, 1051.1 ,  227.7 ,  562.35,  572.7 ,
         1085.6 ],
        [ 427.8 ,  368.  ,  726.8 ,  650.9 ,  419.75, 1062.6 ,  893.55,
          816.5 ],
        [ 416.3 ,  885.5 ,  780.85,  342.7 ,  470.35,  326.6 ,  424.35,
          550.85],
        [1009.7 ,  772.8 ,  849.85,  848.7 ,  741.75,  262.2 ,  657.8 ,
          768.2 ]],

       [[ 892.4 , 1124.7 , 1074.1 ,  692.3 ,  821.1 ,  709.55,  325.45,
          999.

In [29]:
#Convert traffic values into “thousands of vehicles”
traffic/1000

array([[[0.315, 0.175, 0.065, 0.37 , 0.419, 0.173, 0.206, 0.271, 0.463,
         0.419, 0.058, 0.123, 0.306, 0.09 , 0.47 , 0.421, 0.066, 0.406,
         0.289, 0.445, 0.104, 0.394, 0.413, 0.172],
        [0.112, 0.083, 0.25 , 0.406, 0.227, 0.229, 0.104, 0.127, 0.063,
         0.293, 0.459, 0.432, 0.191, 0.398, 0.392, 0.336, 0.208, 0.139,
         0.19 , 0.371, 0.081, 0.363, 0.342, 0.333],
        [0.068, 0.143, 0.127, 0.456, 0.457, 0.144, 0.298, 0.445, 0.206,
         0.124, 0.266, 0.443, 0.065, 0.068, 0.168, 0.121, 0.394, 0.445,
         0.195, 0.48 , 0.185, 0.253, 0.334, 0.339],
        [0.3  , 0.402, 0.094, 0.183, 0.054, 0.121, 0.394, 0.394, 0.356,
         0.414, 0.232, 0.34 , 0.41 , 0.321, 0.394, 0.193, 0.184, 0.296,
         0.135, 0.284, 0.2  , 0.061, 0.162, 0.318],
        [0.27 , 0.112, 0.235, 0.129, 0.414, 0.282, 0.092, 0.287, 0.363,
         0.285, 0.403, 0.356, 0.095, 0.474, 0.251, 0.215, 0.178, 0.452,
         0.073, 0.053, 0.335, 0.322, 0.39 , 0.388],
        [0.162, 0.06

In [30]:
#Normalize temperature between 0 and 1
traffic_norm=(traffic - traffic.min())/(traffic.max()-traffic.min())
traffic_norm

array([[[0.59020045, 0.27839644, 0.03340757, 0.71269488, 0.82182628,
         0.27394209, 0.34743875, 0.4922049 , 0.91982183, 0.82182628,
         0.01781737, 0.16258352, 0.5701559 , 0.08908686, 0.93541203,
         0.82628062, 0.03563474, 0.79287305, 0.53229399, 0.87973274,
         0.12026726, 0.76614699, 0.80846325, 0.27171492],
        [0.13808463, 0.07349666, 0.4454343 , 0.79287305, 0.39420935,
         0.3986637 , 0.12026726, 0.1714922 , 0.02895323, 0.54120267,
         0.91091314, 0.85077951, 0.31403118, 0.77505568, 0.76169265,
         0.63697105, 0.3518931 , 0.19821826, 0.31180401, 0.71492205,
         0.06904232, 0.69710468, 0.65033408, 0.63028953],
        [0.04008909, 0.20712695, 0.1714922 , 0.90423163, 0.9064588 ,
         0.20935412, 0.55233853, 0.87973274, 0.34743875, 0.16481069,
         0.48106904, 0.8752784 , 0.03340757, 0.04008909, 0.26280624,
         0.15812918, 0.76614699, 0.87973274, 0.32293987, 0.95768374,
         0.30066815, 0.45211581, 0.6325167 , 0.64365256]

**Task 12**
- Using masking:
- Replace traffic values < 120 with 0


- Cap sales values above 950 to 950


In [31]:
#Replace traffic values < 120 with 0
np.where(traffic<120,0,traffic)

array([[[315, 175,   0, 370, 419, 173, 206, 271, 463, 419,   0, 123,
         306,   0, 470, 421,   0, 406, 289, 445,   0, 394, 413, 172],
        [  0,   0, 250, 406, 227, 229,   0, 127,   0, 293, 459, 432,
         191, 398, 392, 336, 208, 139, 190, 371,   0, 363, 342, 333],
        [  0, 143, 127, 456, 457, 144, 298, 445, 206, 124, 266, 443,
           0,   0, 168, 121, 394, 445, 195, 480, 185, 253, 334, 339],
        [300, 402,   0, 183,   0, 121, 394, 394, 356, 414, 232, 340,
         410, 321, 394, 193, 184, 296, 135, 284, 200,   0, 162, 318],
        [270,   0, 235, 129, 414, 282,   0, 287, 363, 285, 403, 356,
           0, 474, 251, 215, 178, 452,   0,   0, 335, 322, 390, 388],
        [162,   0,   0, 129, 293, 323, 228,   0, 459, 416, 482, 284,
         323, 466, 387, 283,   0, 140, 318, 336, 259, 451, 194, 178],
        [209, 407, 251, 498, 294, 430, 472, 304,   0, 146, 372,   0,
         496, 145, 405, 461, 132,   0, 296, 127, 413,   0, 423, 287]],

       [[165, 471, 317, 4

In [32]:
#Cap sales values above 950 to 950
np.where(sales>950,950,sales)

array([[[737, 936, 230, 627, 555, 197, 950, 663],
        [681, 264, 814, 128, 949, 411, 300, 490],
        [372, 950, 888, 615, 532, 560, 688, 286],
        [165, 694, 891, 636, 556, 410, 727, 121],
        [794, 759, 451, 410, 103, 723, 710, 509]],

       [[781, 114, 813, 683, 266, 533, 890, 724],
        [709, 147, 259, 914, 198, 489, 498, 944],
        [372, 320, 632, 566, 365, 924, 777, 710],
        [362, 770, 679, 298, 409, 284, 369, 479],
        [878, 672, 739, 738, 645, 228, 572, 668]],

       [[776, 950, 934, 602, 714, 617, 283, 869],
        [813, 535, 241, 619, 609, 894, 622, 600],
        [850, 354, 266, 950, 365, 950, 798, 773],
        [890, 950, 428, 538, 679, 396, 866, 787],
        [843, 870, 208, 635, 337, 950, 515, 933]],

       [[668, 904, 699, 744, 690, 199, 415, 826],
        [386, 715, 872, 827, 393, 950, 867, 934],
        [671, 647, 678, 797, 286, 868, 792, 235],
        [222, 769, 142, 854, 556, 820, 854, 619],
        [260, 184, 368, 692, 857, 714, 200, 

**Task 13 – Product Discount**
- Create:
- discount = np.array([10,20,30,40,50,60,70,80])

- Apply this discount to every store and month using broadcasting


In [33]:
discount = np.array([10,20,30,40,50,60,70,80])
discount

array([10, 20, 30, 40, 50, 60, 70, 80])

In [34]:
sales-discount

array([[[727, 916, 200, 587, 505, 137, 904, 583],
        [671, 244, 784,  88, 899, 351, 230, 410],
        [362, 975, 858, 575, 482, 500, 618, 206],
        [155, 674, 861, 596, 506, 350, 657,  41],
        [784, 739, 421, 370,  53, 663, 640, 429]],

       [[771,  94, 783, 643, 216, 473, 820, 644],
        [699, 127, 229, 874, 148, 429, 428, 864],
        [362, 300, 602, 526, 315, 864, 707, 630],
        [352, 750, 649, 258, 359, 224, 299, 399],
        [868, 652, 709, 698, 595, 168, 502, 588]],

       [[766, 958, 904, 562, 664, 557, 213, 789],
        [803, 515, 211, 579, 559, 834, 552, 520],
        [840, 334, 236, 948, 315, 906, 728, 693],
        [880, 938, 398, 498, 629, 336, 796, 707],
        [833, 850, 178, 595, 287, 933, 445, 853]],

       [[658, 884, 669, 704, 640, 139, 345, 746],
        [376, 695, 842, 787, 343, 918, 797, 854],
        [661, 627, 648, 757, 236, 808, 722, 155],
        [212, 749, 112, 814, 506, 760, 784, 539],
        [250, 164, 338, 652, 807, 654, 130, 

**Task 14 – Add City Tax**
- Add 5% tax to all sales values using broadcasting


In [35]:
sales*1.05

array([[[ 773.85,  982.8 ,  241.5 ,  658.35,  582.75,  206.85, 1022.7 ,
          696.15],
        [ 715.05,  277.2 ,  854.7 ,  134.4 ,  996.45,  431.55,  315.  ,
          514.5 ],
        [ 390.6 , 1044.75,  932.4 ,  645.75,  558.6 ,  588.  ,  722.4 ,
          300.3 ],
        [ 173.25,  728.7 ,  935.55,  667.8 ,  583.8 ,  430.5 ,  763.35,
          127.05],
        [ 833.7 ,  796.95,  473.55,  430.5 ,  108.15,  759.15,  745.5 ,
          534.45]],

       [[ 820.05,  119.7 ,  853.65,  717.15,  279.3 ,  559.65,  934.5 ,
          760.2 ],
        [ 744.45,  154.35,  271.95,  959.7 ,  207.9 ,  513.45,  522.9 ,
          991.2 ],
        [ 390.6 ,  336.  ,  663.6 ,  594.3 ,  383.25,  970.2 ,  815.85,
          745.5 ],
        [ 380.1 ,  808.5 ,  712.95,  312.9 ,  429.45,  298.2 ,  387.45,
          502.95],
        [ 921.9 ,  705.6 ,  775.95,  774.9 ,  677.25,  239.4 ,  600.6 ,
          701.4 ]],

       [[ 814.8 , 1026.9 ,  980.7 ,  632.1 ,  749.7 ,  647.85,  297.15,
          912.

**Task 15 – Temperature Effect**
- If temperature > 35:
- Increase all sales by 2%
- (Use broadcasting and masking together)


In [36]:
temp=temperature[:12]
sales1=sales[:12]

In [37]:
sales1=sales1[temp>35]*1.02

In [38]:
sales1

array([[[ 751.74,  954.72,  234.6 ,  639.54,  566.1 ,  200.94,  993.48,
          676.26],
        [ 694.62,  269.28,  830.28,  130.56,  967.98,  419.22,  306.  ,
          499.8 ],
        [ 379.44, 1014.9 ,  905.76,  627.3 ,  542.64,  571.2 ,  701.76,
          291.72],
        [ 168.3 ,  707.88,  908.82,  648.72,  567.12,  418.2 ,  741.54,
          123.42],
        [ 809.88,  774.18,  460.02,  418.2 ,  105.06,  737.46,  724.2 ,
          519.18]],

       [[ 517.14,  802.74,  617.1 ,  748.68,  303.96,  588.54,  239.7 ,
          162.18],
        [ 260.1 ,  767.04,  603.84,  791.52,  667.08,  927.18,  551.82,
          391.68],
        [ 198.9 ,  508.98,  342.72,  364.14,  780.3 ,  341.7 ,  264.18,
          703.8 ],
        [ 759.9 ,  638.52,  741.54,  399.84,  590.58,  977.16,  395.76,
          393.72],
        [ 618.12,  773.16,  346.8 ,  661.98,  910.86,  457.98,  369.24,
          905.76]]])

**Task 16 – Find Important Values**
- Extract:
- All sales greater than 800


- All traffic values greater than 450


- Days where temperature < 25



In [39]:
#All sales greater than 800
sales[sales>800]

array([936, 974, 814, 949, 995, 888, 891, 813, 890, 914, 944, 924, 878,
       978, 934, 869, 813, 894, 850, 988, 966, 890, 958, 866, 843, 870,
       993, 933, 904, 826, 872, 827, 978, 867, 934, 868, 854, 820, 854,
       857, 892, 951, 947, 809, 925, 967, 858, 931, 802, 907, 923, 964,
       806, 968, 841, 983, 971, 805, 899, 816, 813, 811, 897, 971, 982,
       904, 814, 857, 812, 925, 912, 984, 916, 968, 949, 973, 910, 936,
       845, 811, 887, 835, 882, 836, 859, 848, 909, 958, 893, 888],
      dtype=int32)

In [40]:
#All traffic values greater than 450
traffic[traffic>450]

array([463, 470, 459, 456, 457, 480, 474, 452, 459, 482, 466, 451, 498,
       472, 496, 461, 471, 455, 467, 477, 462, 496, 467, 499, 486, 482,
       496, 466, 484, 461, 454, 473, 482, 465, 463, 456, 465, 472, 465,
       497, 476, 466, 472, 482, 459, 480, 483, 489, 485, 486, 464, 494,
       455, 462, 485, 474, 471, 472, 470, 459, 472, 457, 473, 472, 499,
       477, 452, 481, 452], dtype=int32)

In [41]:
#Days where temperature < 25
temperature[temperature<25]

array([17.74729097, 24.31127611, 20.09455489, 24.45504814, 24.37602704,
       24.94072962, 16.60015522, 24.60582144, 23.85524207, 23.34533122,
       22.97164746, 20.59979979, 24.92048419, 24.53413882, 22.61839872,
       16.60888874, 23.50869627, 16.46843011, 22.2328058 , 23.94401526,
       23.96648653, 21.99882333, 22.41132915, 18.24430309, 24.11363546,
       24.41492512, 23.68560572, 16.12480185, 24.54876513, 23.41593843,
       20.57552726, 19.8390304 , 20.77630162, 24.71264294, 21.99768299,
       22.6578604 , 24.26819685, 17.81127975, 24.3671662 , 24.7816667 ,
       23.83026681, 24.07214152, 21.77791752, 23.82591347, 22.33406215,
       22.84304465, 18.39727368, 22.43167526, 15.48753335, 17.1255446 ,
       23.76440045, 20.90003964, 22.84189835, 24.04001003])

**Task 17**
- Create a new array where:
- Traffic > 400 → 1


- Otherwise → 0




In [42]:
arr=np.where(traffic>400,1,0)
arr

array([[[0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0,
         0, 1, 0],
        [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0],
        [0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0,
         0, 0, 0],
        [0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0],
        [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
         0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
         1, 0, 0],
        [0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1,
         0, 1, 0]],

       [[0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0,
         1, 0, 0],
        [0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 1],
        [1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0,
         0, 1, 0],
        [0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
         0, 0, 1

**Task 18**
- Find:
- Number of high sale transactions (>850)


In [43]:
sales[sales>850]

array([936, 974, 949, 995, 888, 891, 890, 914, 944, 924, 878, 978, 934,
       869, 894, 988, 966, 890, 958, 866, 870, 993, 933, 904, 872, 978,
       867, 934, 868, 854, 854, 857, 892, 951, 947, 925, 967, 858, 931,
       907, 923, 964, 968, 983, 971, 899, 897, 971, 982, 904, 857, 925,
       912, 984, 916, 968, 949, 973, 910, 936, 887, 882, 859, 909, 958,
       893, 888], dtype=int32)

**Task 19**
- Find:
- Sorted order of monthly sales


- Indices of top 3 selling months


- Worst performing store



In [44]:
sales1=sales.sum(axis=(1,2))
sorted_s=np.sort(sales1)
print("Sorted order of monthly sales:\n",sorted_s)
print("Indices of top 3 selling months:\n",sorted_s[-3:])
print("Worst performing store:\n",sorted_s[0])

Sorted order of monthly sales:
 [19346 19628 20190 20291 20725 21047 21356 21947 22418 22551 24623 26942]
Indices of top 3 selling months:
 [22551 24623 26942]
Worst performing store:
 19346


**Task 20**
- Using:
- argmax()
- argmin()
- argsort()

- Find:
- Busiest region


- Least busy hour


In [37]:
#Busiest region
t=np.sum(traffic,axis=(1,2))
s_t=np.argmax(t)
print("Busiest region:\n",s_t)

Busiest region:
 3


In [46]:
#Least busy hour
t=np.sum(traffic,axis=(0,1))
s_t=np.argmin(t)
print("Least busy hour:\n",s_t)

Least busy hour:
 10


**Task 21**
- Using NumPy random:
- Generate random sample of 50 sales values


- Shuffle traffic data along hours axis


- Randomly pick 10 temperature values



In [47]:
arr=np.random.choice(sales.flatten(),size=(50,),replace=False)
arr

array([933, 393, 904, 336, 114, 599, 836, 579, 681, 119, 214, 402, 489,
       147, 228, 866, 978, 899, 931, 163, 907, 276, 845, 888, 738, 692,
       649, 397, 505, 416, 261, 645, 353, 870, 812, 635, 585, 154, 872,
       363, 622, 184, 320, 340, 108, 776, 543, 850, 813, 806], dtype=int32)

**Task 22**
- Find relationship between:
- Average monthly temperature and monthly sales


- Average daily temperature and daily traffic




In [38]:
av_t=np.mean(temperature)
av_s=np.mean(sales,axis=(1,2))#pendingggg question
print("Average monthly temperature and monthly sales:\n",av_s)

Average monthly temperature and monthly sales:
 [563.775 560.45  673.55  615.575 504.75  533.9   483.65  518.125 526.175
 507.275 490.7   548.675]


**Task 23 – Trend Analysis**
- Compare:
- First 6 months vs last 6 months sales
- Which half of the year performed better?




In [49]:
sales.shape

(12, 5, 8)

In [50]:
print("First 6 months:\n",sales[:6,:,:])

First 6 months:
 [[[737 936 230 627 555 197 974 663]
  [681 264 814 128 949 411 300 490]
  [372 995 888 615 532 560 688 286]
  [165 694 891 636 556 410 727 121]
  [794 759 451 410 103 723 710 509]]

 [[781 114 813 683 266 533 890 724]
  [709 147 259 914 198 489 498 944]
  [372 320 632 566 365 924 777 710]
  [362 770 679 298 409 284 369 479]
  [878 672 739 738 645 228 572 668]]

 [[776 978 934 602 714 617 283 869]
  [813 535 241 619 609 894 622 600]
  [850 354 266 988 365 966 798 773]
  [890 958 428 538 679 396 866 787]
  [843 870 208 635 337 993 515 933]]

 [[668 904 699 744 690 199 415 826]
  [386 715 872 827 393 978 867 934]
  [671 647 678 797 286 868 792 235]
  [222 769 142 854 556 820 854 619]
  [260 184 368 692 857 714 200 421]]

 [[649 684 337 523 642 395 681 267]
  [708 285 892 363 951 585 397 119]
  [731 115 314 216 261 354 457 166]
  [557 498 393 764 605 183 119 163]
  [947 809 451 729 573 735 925 647]]

 [[194 967 858 422 351 360 739 580]
  [931 752 702 158 558 352 802 270]
 

In [51]:
print("First 6 months:\n",sales[6:,:,:])

First 6 months:
 [[[660 587 217 576 394 806 544 730]
  [416 453 292 968 841 166 108 983]
  [177 794 971 121 805 568 412 340]
  [899 651 142 728 133 348 114 335]
  [176 508 569 437 532 109 256 480]]

 [[646 816 813 775 697 811 237 897]
  [112 105 971 684 796 982 649 393]
  [397 346 392 362 377 149 165 281]
  [415 486 529 709 584 321 214 276]
  [402 316 904 742 557 460 565 392]]

 [[256 787 631 358 727 548 551 814]
  [857 263 409 235 812 925 192 674]
  [287 912 666 123 746 238 984 297]
  [505 385 157 756 574 200 338 247]
  [730 331 916 119 680 968 608 241]]

 [[249 783 723 949 578 973 295 910]
  [408 316 782 936 457 477 232 442]
  [663 153 515 306 427 485 337 626]
  [845 811 145 543 482 534 216 730]
  [410 604 393 615 293 376 110 162]]

 [[711 154 887 486 759 835 553 588]
  [343 685 199 350 216 613 449 882]
  [836 108 342 188 744 568 493 687]
  [305 386 349 790 353 402 179 510]
  [859 682 283 240 250 848 315 201]]

 [[507 787 605 734 298 577 235 159]
  [255 752 592 776 654 909 541 384]
 

In [52]:
#Which half of the year performed better?
s1=sales[:6,:,:]
s2=sales[6:,:,:]
np.where(s1>s2,s1,s2)

array([[[737, 936, 230, 627, 555, 806, 974, 730],
        [681, 453, 814, 968, 949, 411, 300, 983],
        [372, 995, 971, 615, 805, 568, 688, 340],
        [899, 694, 891, 728, 556, 410, 727, 335],
        [794, 759, 569, 437, 532, 723, 710, 509]],

       [[781, 816, 813, 775, 697, 811, 890, 897],
        [709, 147, 971, 914, 796, 982, 649, 944],
        [397, 346, 632, 566, 377, 924, 777, 710],
        [415, 770, 679, 709, 584, 321, 369, 479],
        [878, 672, 904, 742, 645, 460, 572, 668]],

       [[776, 978, 934, 602, 727, 617, 551, 869],
        [857, 535, 409, 619, 812, 925, 622, 674],
        [850, 912, 666, 988, 746, 966, 984, 773],
        [890, 958, 428, 756, 679, 396, 866, 787],
        [843, 870, 916, 635, 680, 993, 608, 933]],

       [[668, 904, 723, 949, 690, 973, 415, 910],
        [408, 715, 872, 936, 457, 978, 867, 934],
        [671, 647, 678, 797, 427, 868, 792, 626],
        [845, 811, 145, 854, 556, 820, 854, 730],
        [410, 604, 393, 692, 857, 714, 200, 

**Task 24 – Growth Rate**
- Calculate:
- Month-to-month percentage growth in sales


In [39]:
sales.shape

(12, 5, 8)

In [42]:
monthly_sales = sales.sum(axis=(1,2))  
growth_rate = (monthly_sales[1:] - monthly_sales[:-1]) / monthly_sales[:-1] * 100
print("Month-to-month percentage growth in sales:\n",growth_rate)

Month-to-month percentage growth in sales:
 [ -0.58977429  20.18021233  -8.60737881 -18.00349267   5.77513621
  -9.41187488   7.12808849   1.55367913  -3.59196085  -3.26745848
  11.81475443]


**Task 25**
- Find:
- For each region → which hour has highest average traffic


